src/local_scheduler/local_scheduler_algorithm.cc

```diff
@@ -137,18 +148,15 @@ SchedulingAlgorithmState *SchedulingAlgorithmState_init(void) {
 
 void SchedulingAlgorithmState_free(SchedulingAlgorithmState *algorithm_state) {
   /* Free all of the tasks in the waiting queue. */
-  task_queue_entry *elt, *tmp1;
-  DL_FOREACH_SAFE(algorithm_state->waiting_task_queue, elt, tmp1) {
-    DL_DELETE(algorithm_state->waiting_task_queue, elt);
-    free(elt->spec);
-    free(elt);
+  for (auto &task : *algorithm_state->waiting_task_queue) {
+    TaskQueueEntry_free(&task);
```

Comment:
(Or maybe just `std::list<TaskQueueEntry>` -- depends on what you're trying to do.)